packages/react-native/React/Views/RCTViewManager.m

```diff
@@ -292,34 +291,45 @@ - (RCTShadowView *)shadowView
 
 RCT_CUSTOM_VIEW_PROPERTY(accessibilityRole, UIAccessibilityTraits, RCTView)
 {
-  #if !TARGET_OS_OSX // [macOS]
+#if !TARGET_OS_OSX // [macOS]
   UIAccessibilityTraits accessibilityRoleTraits =
       json ? [RCTConvert UIAccessibilityTraits:json] : UIAccessibilityTraitNone;
   if (view.reactAccessibilityElement.accessibilityRoleTraits != accessibilityRoleTraits) {
     view.accessibilityRoleTraits = accessibilityRoleTraits;
     view.reactAccessibilityElement.accessibilityRole = json ? [RCTConvert NSString:json] : nil;
     [self updateAccessibilityTraitsForRole:view withDefaultView:defaultView];
   }
-  #else // [macOS
-    if (json) {
-      view.reactAccessibilityElement.accessibilityRole = [RCTConvert accessibilityRoleFromTraits:json];
-    } else {
-      view.reactAccessibilityElement.accessibilityRole = defaultView.accessibilityRole;
-    }
-  #endif // macOS]
+#else // [macOS
+  // accessibilityRoleInternal is used to cache the converted value from the prop
+  view.reactAccessibilityElement.accessibilityRoleInternal = json ? [RCTConvert accessibilityRoleFromTraits:json usingAriaMappings:NO] : nil;
+  // update the actual NSAccessibilityRole if it doesn't match
+  if (view.reactAccessibilityElement.accessibilityRole != view.reactAccessibilityElement.accessibilityRoleInternal) {
+    [self updateAccessibilityRole:view withDefaultView:defaultView];
+  }
+#endif // macOS]
 }
 
-#if !TARGET_OS_OSX // [macOS]
+
```

Comment:
Nit: Do we want the extra space? Asking because during merges, I have spent time removing random extra spaces we've introduced 😅